system-helper/flatpak-system-helper.c

```diff
@@ -217,6 +296,26 @@ flatpak_invocation_return_error (GDBusMethodInvocation *invocation,
     }
 }
 
+static OngoingPull *
+take_ongoing_pull_by_dir (const gchar *src_dir)
+{
+  OngoingPull *pull = NULL;
+
+  G_LOCK (ongoing_pulls);
+  for (guint i = 0; i < ongoing_pulls->len; i++)
+    {
+      pull = g_ptr_array_index (ongoing_pulls, i);
```

Comment:
This seems wrong to me. If we're not hitting any ongoing pull, we will still return the last object in the array. You need to use a different variable here and only assign the pull variable we return from the function if the strcmp succeeds.

Reply:
Yes, this is way wrong.

Re-worked with the correct logic.